extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java

```diff
@@ -297,29 +297,55 @@ void registerClasses(
     }
 
     @BuildStep
-    @Record(ExecutionTime.STATIC_INIT)
-    void setupJaxbContextConfig(JaxbConfig config,
-            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems,
-            JaxbContextConfigRecorder jaxbContextConfig) {
-        Set<String> classNamesToBeBound = new HashSet<>();
-        for (JaxbClassesToBeBoundBuildItem classesToBeBoundBuildItem : classesToBeBoundBuildItems) {
-            classNamesToBeBound.addAll(classesToBeBoundBuildItem.getClasses());
-        }
+    FilteredJaxbClassesToBeBoundBuildItem filterBoundClasses(
+            JaxbConfig config,
+            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems) {
+
+        FilteredJaxbClassesToBeBoundBuildItem.Builder builder = FilteredJaxbClassesToBeBoundBuildItem.builder();
+        classesToBeBoundBuildItems.stream()
+                .map(JaxbClassesToBeBoundBuildItem::getClasses)
+                .forEach(builder::classNames);
 
         // remove classes that have been excluded by users
         if (config.excludeClasses.isPresent()) {
-            classNamesToBeBound.removeAll(config.excludeClasses.get());
+            builder.classNameExcludes(config.excludeClasses.get());
         }
+        return builder.build();
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void setupJaxbContextConfig(
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            JaxbContextConfigRecorder jaxbContextConfig) {
+        jaxbContextConfig.addClassesToBeBound(filteredClassesToBeBound.getClasses());
+    }
 
-        // parse class names to class
-        Set<Class<?>> classes = getAllClassesFromClassNames(classNamesToBeBound);
-        if (config.validateJaxbContext) {
-            // validate the context to fail at build time if it's not valid
-            validateContext(classes);
+    @BuildStep
+    void validateDefaultJaxbContext(
+            JaxbConfig config,
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            SynthesisFinishedBuildItem beanContainerState,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> errors) {
+
+        final BeanResolver beanResolver = beanContainerState.getBeanResolver();
+        final Set<BeanInfo> beans = beanResolver
+                .resolveBeans(Type.create(DotName.createSimple(JAXBContext.class), org.jboss.jandex.Type.Kind.CLASS));
```

Comment:
Missing the Marshaller and Unmarshaller types too.

Reply:
Marshaller and Unmarshaller require the JAXBContext bean to exist, so checking their existence is not necessary IMO.  Let me add a test to prove that.

Reply:
There is now ConflictingModelClassesMarshalerOnlyTest to cover the situation where there conflicting classes, no JAXBContext injection point, but there is a Marshaller injection point. The application fails on boot as expected.